File: python/1619_restaurant_customers.py
n = int(input())

arrivals = []
departures = []
for _ in range(n):
    a, d = map(int, input().split())
    arrivals.append(a)
    departures.append(d)

arrivals.sort()
departures.sort()

i, j = 0, 0

m = 0
s = 0
while i < n and j < n:
    if arrivals[i] < departures[j]:
        i += 1
        s += 1
    elif arrivals[i] > departures[j]:
        j += 1
        s -= 1
    else:
        i += 1
        j += 1

    m = max(s, m)

print(m)

File: src/1619_restaurant_customers.rs
use std::cmp::Ordering;
use std::io::{self, BufWriter, Write};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let input = io::stdin();
    let mut lines = input.lines();
    // Flush all at once
    let mut output = BufWriter::new(io::stdout().lock());

    let n = lines.next().unwrap().unwrap().parse::<usize>()?;

    let mut arrivals = Vec::with_capacity(n);
    let mut departures = Vec::with_capacity(n);

    for line in lines {
        let line = line?;
        let (a, d) = line.split_once(' ').unwrap();
        let a = a.parse::<i64>()?;
        let d = d.parse::<i64>()?;
        arrivals.push(a);
        departures.push(d);
    }

    arrivals.sort_unstable();
    departures.sort_unstable();

    let (mut state, mut max) = (0, 0);
    let (mut i, mut j) = (0, 0);

    while i < n && j < n {
        match arrivals[i].cmp(&departures[j]) {
            Ordering::Less => {
                i += 1;
                state += 1;
            }
            Ordering::Equal => {
                i += 1;
                j += 1;
            }
            Ordering::Greater => {
                j += 1;
                state -= 1;
            }
        }
        max = max.max(state);
    }

    writeln!(output, "{max}")?;

    Ok(())
}
